Add cap verify command

toniecrumley opened the pull request:
## Summary
<!-- Describe your science -->
The `cap verify` command facilitates verifying the reproducibility of a pipeline.  It provides conventions for where to place verification scripts and their output, as well as helper functions for generating verification output files.  When the script and corresponding output files are committed to source control, reviewers and other scientists interested in the work can deterministically verify the reproducibility of our work.

Verification files and their output are stored in the `verifications` folder in the root of a project.  The verification files are shell scripts ending in `.sh` and the output is written to the same file name with a `.out` extension.  The verification file `verifications/example.sh` would create the output file `verifications/example.out`.

## Author Checklist
- [x] This repository has a Lasseigne Lab ruleset configured.
  See instructions [here](https://github.com/lasseignelab/project-template/blob/main/DEVELOPMENT.md).
- [x] Only files relevant to this change are included; no temporary, generated,
  or unrelated files are part of this pull request. Use `.gitignore` to prevent
  irrelevant files from accidentally being included.
- [x] All automated tests have passed.
- [x] All automated code checks have passed. The "All checks have passed"
  message should show in the GitHub pull request status box above the
  "Merge pull request" button.
- [x] No merge conflicts exist. Merge conflicts are indicated by the "This
  branch has conflicts that must be resolved" message in the GitHub pull
  request status box above the "Merge pull request" button.
- [x] Has the CHANGELOG been updated with the planned version number?
- [x] Setup instructions have been provided.
- [x] Testing instructions have been provided.
- [x] Primary and secondary reviewers have been requested to review this
  pull request.

## Self/Peer Review Checklist ([Coding Guidelines](https://docs.google.com/document/d/1h1hxQGrqnQqo1pAxrrtX1OtjHddRTqtgsgDFSXOQpzk/edit?usp=sharing))
- Meaningful variable and function names
- File header comments
- Function comments
- In-line comments summarize logical sections of code by concisely explaining
  why, not what the code is doing.  Avoid excessive or redundant commenting.
- Confirm that the automated tests pass.
- Confirm the code performs the intended functionality

## Setup
<!-- How to setup the project before testing the functionality? -->
Get the code for the pull request.
```
cd ~/bin/capture
git checkout main
git pull
git checkout cap-verify
source cap_completion.sh
cd project-template
git checkout main
git pull
git checkout version-1-0
cd ..

```

## Testing the functionality
<!-- How to test the functionality? -->
Run the automated test suite.
```
cd ~/bin/capture
tests/install
tests/run

```
Test manually in scratch:
```
cd $USER_SCRATCH
cap new --owner lasseignelab "$USER-testing"

```
```
cd "$USER-testing"

```
Test verifications manually in a sandbox project or a project you are currently working on:
* Create a verification with a call to `cap_verify_md5`.
* Create a verification with a call to `cap_verify_append`.
* Create a verification with multiple `cap_verify_md5` and `cap_verify_append` functions.
* Execute the verifications created with various combinations of the `cap verify` command options.
* Execute verifications that use `cap_verify_md5` with various combinations of `cap_verify_md5` options.

## Cleanup
* Remove the test directory from scratch:
```
cd "$USER_SCRATCH"
rm -rf "$USER-testing"

```
* Delete the test repo from github.
* Reset CAPTURE to the released version.
```
cap update

```

toniecrumley committed:
Add cap_md5_verify command

Initial version of the command with tests but not fully integrated
with the cap verify command.

toniecrumley committed:
Merge branch 'main' into cap-verify

Conflicts:
	README.md
	lib/environment.sh

toniecrumley committed:
Merge branch 'main' into cap-verify

toniecrumley committed:
Merge branch 'main' into cap-verify

toniecrumley committed:
Add cap_verify_md5 to the README.md

toniecrumley committed:
Create a containerized test runner

Changes to my HPC environment broke npm.  In order to insulate the tests
from HPC changes, a test runner was created to use a docker container to
run the test suite.

toniecrumley committed:
Change verification output file extension

toniecrumley committed:
Fix master branch issue with new git

Old git versions would name the default branch master. This is an
insensitive name so new versions of git use main as the default.
If a user has the old version of git we rename name master to main.
If they have a new version we get an error attempting to rename a
master branch that does not exist. To resolve the issue a check for
master existing needs to be added.

toniecrumley committed:
Fix jq missing in containerized tests

The jq, JSON query, command is not installed everywhere so it did
not work in the container used for testing. It also would not work
in a lot of linux installations. Therefore, it is being replaced by
grep and cut which are ubiquitous.

toniecrumley committed:
Add test framework install script

toniecrumley committed:
Document running the test suite

toniecrumley committed:
Add Contributions to the README table of contents

toniecrumley committed:
Add --output-files-only option to cap md5

Output for CAPTURE verifications should be reproducable. As a result,
output that only serves as human readable help needed to be removed.
This commit adds an option for removing informational text so that
cap verify and cap_verify_md5 output will be clean and reproducible.

toniecrumley committed:
Add the cap_verify_md5 verifications helper function

toniecrumley committed:
Update cap verify documentation

toniecrumley committed:
Link to verification helper functions

toniecrumley committed:
Fix link to verification helper functions

toniecrumley committed:
Fix documentation formatting

toniecrumley committed:
Add cap verify command

toniecrumley committed:
Append cap_verify_md5 output to facilitate multiple calls

toniecrumley committed:
Move Slurm job creation from cap_verify_md5 to cap verify

toniecrumley committed:
Add cap_verify_append function

toniecrumley committed:
Set CAP_VERIFICATION_DRY_RUN explicitly to false

toniecrumley committed:
Document cap verify environment variables

toniecrumley committed:
Fix cap_verify_md5 documentation

toniecrumley committed:
Improve example documentation

toniecrumley committed:
Fix verifications in subdirectories not saving correctly

toniecrumley committed:
Merge branch 'main' into cap-verify

toniecrumley committed:
Update the CHANGELOG

toniecrumley committed:
Improve the contributor portion of the README

toniecrumley committed:
Fix test broken by merging main

toniecrumley committed:
Make cap verify documentation and help consistent

toniecrumley committed:
Fix cap verify documentation and help errors

toniecrumley committed:
Fix linter errors

toniecrumley committed:
Upgrade megalinter to version 9

toniecrumley committed:
Fix path issue for tests on github

toniecrumley committed:
Fix install path issue on Github

tsoelter commented on README.md:
```diff
@@ -286,8 +299,10 @@ Submitted batch job 29818073
-- **CAP_PROJECT_NAME**: The name of the project given with the `cap new`
-command.
+- **CAP_CONTAINER_PATH**: Path to where container files such as Docker will be
+maintained.  Defaults to `<project-path>/bin/docker`.
```
Shouldn't this say `<project-path>/bin/container`, since it replaced `<project-path>/bin/docker`?

toniecrumley replied:
Yes, this change has been made.

tsoelter commented on README.md:
```diff
@@ -296,19 +311,19 @@ be set in the `~/.caprc` file.  It is possible to set it as a shell environment
+- **CAP_RESULTS_PATH**: Path to where analysis results will be written.
+Defaults to `<project-path>/results`.
+- **CAP_VERIFICATIONS_PATH**: Path to where verification files and the
+result files they produce are written.
```
You added a "Defaults to `<project-path>/rest/of/path`" to most of these. Should we add "Defaults to `<project-path>/verifications`?

toniecrumley replied:
Yes, this change has been made.

tsoelter reviewed (commented):
I have tried many different combinations of the helper functions and their arguments and always got what I expected. I also tested the cap verify arguments. Dry run works great, but for some reason `--slurm=batch` works, but `-s=batch` does not? The same applies to `-s=run`. I will include my error below. I don't know if this is unique to cap verify.

```
Error: invalid value for --slurm option
Use the 'cap help verify' command for detailed help.
```

toniecrumley commented:
> I have tried many different combinations of the helper functions and their arguments and always got what I expected. I also tested the cap verify arguments. Dry run works great, but for some reason `--slurm=batch` works, but `-s=batch` does not? The same applies to `-s=run`. I will include my error below. I don't know if this is unique to cap verify.
> 
> ```
> Error: invalid value for --slurm option
> Use the 'cap help verify' command for detailed help.
> ```

@tsoelter That is the way commandline parsers work.  Below is a summary table from ChatGPT.

<img width="781" height="292" alt="Screenshot 2025-12-16 at 4 14 09 PM" src="https://github.com/user-attachments/assets/cc43ec67-cbad-49ba-8fb0-6f845560aa09" />

tsoelter commented:
> > I have tried many different combinations of the helper functions and their arguments and always got what I expected. I also tested the cap verify arguments. Dry run works great, but for some reason `--slurm=batch` works, but `-s=batch` does not? The same applies to `-s=run`. I will include my error below. I don't know if this is unique to cap verify.
> > ```
> > Error: invalid value for --slurm option
> > Use the 'cap help verify' command for detailed help.
> > ```
> 
> @tsoelter That is the way commandline parsers work. Below is a summary table from ChatGPT.
> 
> <img alt="Screenshot 2025-12-16 at 4 14 09 PM" width="781" height="292" src="https://private-user-images.githubusercontent.com/148514625/527296209-cc43ec67-cbad-49ba-8fb0-6f845560aa09.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjU5MjUxMTMsIm5iZiI6MTc2NTkyNDgxMywicGF0aCI6Ii8xNDg1MTQ2MjUvNTI3Mjk2MjA5LWNjNDNlYzY3LWNiYWQtNDliYS04ZmIwLTZmODQ1NTYwYWEwOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMjE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTIxNlQyMjQwMTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02NmMxZmEwNDMwZDMyNzIwYjRkZDc2YTVlNWVjOGJkMzlhNWJjNDEyYzcxYjczZjZmYWU3ZDc5YjM0OTFlNTdjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.m1ugqTCEKfqkd0Kr1KUxeGAzPz7ztuPiRyQYxtO1Y74">

Ahh, not sure how I never noticed that before. 🤦‍♀️  Thanks. Please disregard my comment then, @toniecrumley.

toniecrumley committed:
Fix README errors

toniecrumley committed:
Require commands run from project root directory

The env, run, and verify commands must run from the project root
directory for everything to function properly. If they are ran outside
of a project root directory then an error message will be displayed.

toniecrumley committed:
Fix linter issues

tchowton commented on commands/verify.sh:
```diff
@@ -0,0 +1,178 @@
+      ;;
+  esac
+
+  # Output to the logs folder like the cap run command.
```
I think you can remove this line

toniecrumley replied:
Agreed.  It has been removed.

toniecrumley committed:
Remove unneeded comment

toniecrumley committed:
Replace conda with env

toniecrumley committed:
Change bin/conda to bin/env

To generalize functionality related to application runtime environment,
the conda specific code is replaced by generic environment code.

Since CAPTURE has the concept of config/environments, the decision
was made to change bin/conda to bin/env and CAP_CONDA_PATH to
CAP_ENV_PATH. To reduce confusion around config/environments, the
CAP_ENV environment variable was renamed CAP_ENVIRONMENT.

toniecrumley committed:
Update the CHANGELOG

toniecrumley committed:
Merge pull request #52 from lasseignelab/rename-conda

Rename bin/conda to bin/env

toniecrumley committed:
Merge branch 'cap-verify' into cap-run-from-root

toniecrumley committed:
Merge pull request #48 from lasseignelab/cap-run-from-root

Require commands run from project root directory

tchowton reviewed (approved):
Everything I tested worked properly. However, I think we need to add tab completion for cap verify in a different pull request.

toniecrumley commented:
> Everything I tested worked properly. However, I think we need to add tab completion for cap verify in a different pull request.

@tchowton The tab completion is there but it is loaded in `~/.bash_profile` so it will be old until the completion file is sourced.  I added that to the directions.

```
source cap_completion.sh
```

https://github.com/lasseignelab/capture/pull/13/files#diff-46bb93671b1e09f0de97da70ad43bd169860cbf5af2879d6f62671de37c26c68R8
